### PR TITLE
Improve datalayer

### PR DESCRIPTION
**Pull request by dylan-provost**

### CODE REVIEW

File: w3cdatalayer/application/factory.go [hunk 1]
```diff
@@ -302,6 +311,10 @@ func (s Factory) getProductCategory(product productDomain.BasicProduct) *domain.
 func (s Factory) getProductInfo(product productDomain.BasicProduct) domain.ProductInfo {
 	baseData := product.BaseData()
 	retailerCode := baseData.RetailerCode
+
+	productName := baseData.Title
+	productName = s.regex.ReplaceAllString(baseData.Title, "-")
```
[the remark this answers is not in the document]
Comment (dylan-provost) [reply w3cdatalayer/application/factory.go]: It used to be private and was then created in the inject on suggestion from Thomas Neumann for potential performance issues...

File: w3cdatalayer/application/factory.go [hunk 2]
```diff
@@ -65,6 +67,13 @@ func (s *Factory) Inject(
 	s.locale = config.Locale
 	s.defaultCurrency = config.DefaultCurrency
 	s.hashUserValues = config.HashUserValues
+
+	regexString := "[,|;|\\|]"
```
[the remark this answers is not in the document]
Comment (dylan-provost) [reply w3cdatalayer/application/factory.go]: This applies only to certain rules for Adobe data catcher on product names. We could escape more but it was not a requirement in the story (https://targetprocess.aoe.com/entity/194598-26-change-datalayer-data). Would you suggest to extend i?

File: w3cdatalayer/application/factory.go [hunk 3]
```diff
@@ -65,6 +67,13 @@ func (s *Factory) Inject(
...
+	r, err := regexp.Compile(regexString)
+	if err != nil {
+		panic(err)
```
Comment (chetan-thapliyal) [inline w3cdatalayer/application/factory.go]: Do we really need a panic here? It'll ultimately result in crashing the whole app. Is this really the desired behavior?

File: w3cdatalayer/application/factory.go [hunk 4]
```diff
@@ -302,6 +311,10 @@ func (s Factory) getProductCategory(product productDomain.BasicProduct) *domain.
 func (s Factory) getProductInfo(product productDomain.BasicProduct) domain.ProductInfo {
 	baseData := product.BaseData()
 	retailerCode := baseData.RetailerCode
+
+	productName := baseData.Title
```
Comment (chetan-thapliyal) [inline w3cdatalayer/application/factory.go]: This statement is redundant. productName is immediately overwritten in next statement.

File: w3cdatalayer/application/factory.go (hunk 3 above)
Comment (dylan-provost) [reply w3cdatalayer/application/factory.go]: If that can answer your question, see the discussion with Thomas here https://gitlab.aoe.com/shared/i-love-flamingo/flamingo-commerce/merge_requests/58#note_44273

Comment (masmrlar) [reply w3cdatalayer/application/factory.go]: @chetan-thapliyal I agree with you that `panic` should always be avoided. 
In this case we got a pretty good reason for the panic: If we have an erratic regex string, the regex won't compile and we won't get the expected functionality, but lots of possible error logs. 

So a better way would be to just panic at this point - as an erratic regex is equivalent to broken code.

File: w3cdatalayer/application/factory.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dylan-provost) [reply w3cdatalayer/application/factory.go]: @danielpoe is this still an open question?

**Pull request merged**